composer/datasets/c4.py

```diff
@@ -0,0 +1,333 @@
+    """
+
+    def __init__(self,
+                 split,
+                 max_samples,
+                 tokenizer_name,
+                 max_seq_len,
+                 group_method,
+                 shuffle=False,
+                 shuffle_buffer_size=10000,
+                 seed=5):
+        try:
+            import datasets
+            import transformers
+        except ImportError:
+            raise ImportError('HuggingFace transformers and datasets are not installed. '
+                              'Please install with `pip install composer[nlp]`')
+
+        self.split = split
+        self.max_samples = max_samples
+        self.tokenizer_name = tokenizer_name
+        self.max_seq_len = max_seq_len
+        self.group_method = group_method
+        self.shuffle = shuffle
+        self.shuffle_buffer_size = shuffle_buffer_size
+        self.seed = seed
+
+        # Metadata
+        if self.split == "train":
+            self.num_shards = 1024
+            self.approx_samples_per_shard = 356317
+        elif self.split == "validation":
+            self.num_shards = 8
+            self.approx_samples_per_shard = 45576
+        else:
+            raise ValueError(f"Unknown {split=}, expected 'train' or 'validation'.")
+
+        # Set dataset size
+        self.world_size = dist.get_world_size()
+        self.rank = dist.get_global_rank()
+        self.max_samples_per_device = self.max_samples // self.world_size
+        new_max_samples = self.max_samples_per_device * self.world_size
+        if new_max_samples != self.max_samples:
+            log.warning(
+                f"Max samples will be truncated from {max_samples}->{new_max_samples} to maintain divisibility across {self.world_size} devices."
+            )
+            self.max_samples = new_max_samples
+
+        # Try and detect if max_samples is too large
+        # TODO: For safety, repeat original dataset if max_samples is larger than original size
+        original_approx_samples = self.num_shards * self.approx_samples_per_shard
+        if self.max_samples > original_approx_samples and self.group_method == "truncate":
+            raise ValueError(
+                f"Max samples was set to {self.max_samples} with group_method 'truncate' but split '{split}' has only {original_approx_samples}.)"
+            )
+        if self.group_method == "concat":
+            log.warning(
+                f"When using group_method 'concat', sequential token samples are concatenated and chunked into fixed-length samples of size {self.max_seq_len=}. "
+                f"In general we cannot detect ahead-of-time if your setting of {self.max_samples=} is safe, "
+                f"so please be cautious to not request more tokens than the size of the original dataset.")
+
+        # Build tokenizer
+        self.tokenizer = transformers.AutoTokenizer.from_pretrained(tokenizer_name)
+        if self.tokenizer_name in ["gpt2"]:
```

Comment:
Can we just check `if self.tokenizer.pad_token_id is None` rather than checking if the tokenizer name matches a very specific tokenizer?

Reply:
Yup! As long as you think that the EOS -> PAD replacement is generally safe, and not just a GPT2 thing.